Getting Register out of the sourcecode.

krosk93 opened the pull request:
Register should be used as a plugin, Nijikokun said that recently, and
I think it's the best way too. That way Register can be easily updated,
and we avoid duplicated code.

krosk93 committed:
Using Register as external jar

Register should be used as a plugin, Nijikokun said that recently, and
I think it's the best way too. That way Register can be easily updated,
and we avoid duplicated code.

khobbits commented:
Don't think this would work very well, remember essentials itself is a register plugin, our 'blend' of register needs to remove essentials eco, otherwise we cause a rather annoying loop.

Also without register, essentials would conflict with iconomy and such like, for the same reason pointed above.  

While we are embedding register, we're embedding it correctly, using our own namespace, and thus it wont cause conflicts, doesn't require an extra dependency, and actually means essentials can be used as an economy bridge to other eco plugins.

krosk93 commented:
It's working fine:
2011-10-23 08:32:04 [INFO] [Register] Payment method found (Essentials version: 2.2)
And loading it as an external jar doesn't mean that economy can't be used as a bridge to other eco plugins (but anyway, as Essentials internally uses Register, and we can use it directly, why the hell would anybody use Essentials in that way?)
No annoying loops have been created.

But there are some points in my code that should be fixed: plugin.yml should include Register as a dependency, and should have an autodownloader, I'll make a new commit soon fixing this.

krosk93 committed:
Autodownload Register

Nearly all the autodownload code was made by NarrowTux (Moritz
Schmale). All credits to him.
But I have found that, while autodownload works, errors happened while
trying to enable the plugin directly from the code. The only solution I
have found for it is to shutdown the server, telling the user to reopen
it. I expect someone will find a better way to do it.

The code that did the errors can be checked here:
https://gist.github.com/1307028

krosk93 commented:
There we go with the autodownloader. It had some bugs though, and to avoid them, as soon as the download finishes, it stops the server asking the admin to start it again. The bug was that it cannot load the plugin after downloading it, and I do not know why. I have made a gist with the code I was using to load the plugin after download, it's available here: https://gist.github.com/1307028

khobbits commented:
I pulled your changes, and added a logger line to the get money call in the eco api, then typed /balance.

https://gist.github.com/c568a9791e007b77fe2e

When any external or internal command, calls the ess eco api, it does a little record keeping, and then passes the call down the line, where the register hooks are, this means that, if you don't have any other eco plugin installed, it goes like this.

Command call -> ess eco api -> register -> ess eco api -> register .....

snowleo commented:
Register is included for a reason, as stated by Khobbits above.

We won't pull this request.

khobbits commented:
The rule with EssentialsEco is that "If there is any other available economy, use it."
There is no way with the current register, to tell it to do so reliably, and no way of telling register "I'm sorry that method wont do".

With the current register, if no preferred method is found, it just uses the first one available, which could be Essentials.  So even disabling register internally, would cause problems.
